### PR TITLE
Log "can/should run" and caching in dispatch machinery

### PR DESCRIPTION
**Pull request by eriknw**
This adds informative logging messages for `can_run` and `should_run` to better understand behaviors of dispatching.

### CODE REVIEW

**Comment by eriknw** [comment]
Also added logging for using the cache. This continues logging that was first added in #7300. CC @rlratzel

**Review by Schefflera-Arboricola — COMMENTED** [review]
Thanks @eriknw :)

File: networkx/utils/backends.py [hunk 1]
```diff
@@ -1314,8 +1375,11 @@ def _convert_and_call(self, backend_name, args, kwargs, *, fallback_to_nx=False)
                 backend_name, args, kwargs, use_cache=config.cache_converted_graphs
             )
             _logger.debug(
-                f"using backend '{backend_name}' for call to `{self.name}' "
-                f"with args: {converted_args}, kwargs: {converted_kwargs}"
+                "Using backend '%s' for call to `%s' with args: %s, kwargs: %s",
+                backend_name,
+                self.name,
+                converted_args,
+                converted_kwargs,
```
[the remark this answers is not in the document]
Comment (eriknw) [reply networkx/utils/backends.py:1366]: This is a good thought. How about adding a new log message _when_ falling back to networkx? `fallback_to_nx` isn't exposed to end-users atm--it's currently only an option when testing backends.

File: networkx/utils/backends.py [hunk 2]
```diff
@@ -951,25 +954,62 @@ def _can_backend_run(self, backend_name, /, *args, **kwargs):
...
-        return (
-            hasattr(backend, self.name)
-            and (can_run := backend.can_run(self.name, args, kwargs))
-            and not isinstance(can_run, str)
-        )
+        # why they can't or shouldn't be run.
+        if not hasattr(backend, self.name):
+            _logger.debug(
+                "Backend '%s' does not implement `%s'", backend_name, self.name
+            )
```
[the remark this answers is not in the document]
Comment (eriknw) [reply networkx/utils/backends.py]: Hmm, that duplication is kind of unfortunate. Nice attention to detail.

I just pushed a change that doesn't write that log message when that exception will be raised. The log message may have info that doesn't show up in the warnings--specifically, the string returned by `can_run`--so maybe a little duplication is okay if the log message may be more helpful.

File: networkx/utils/backends.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eriknw) [reply networkx/utils/backends.py:1377]: Done: we now log when falling back to networkx :+1:

**Review by rlratzel — APPROVED** [review]

**Review by Schefflera-Arboricola — APPROVED** [review]
Thank you @eriknw and @rlratzel !

**Review by MridulS — APPROVED** [review]
Not a blocker for this PR but we really need to find good ways to test the backends.py file.

**Pull request merged**